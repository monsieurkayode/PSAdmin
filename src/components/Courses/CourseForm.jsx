import React from 'react';
import {
  string, shape, bool, arrayOf, func
} from 'prop-types';
import TextInput from '../common/TextInput/Index';
import SelectInput from '../common/SelectInput/Index';

/**
 * @function CourseForm
 *
 * @param {object} props
 *
 * @returns {JSX}
 */
const CourseForm = ({
  title,
  authorId,
  category,
  length,
  errors,
  saving,
  loading,
  authors,
  onChange,
  onSubmit,
  onFocus
}) => (
  <form onSubmit={onSubmit}>
    <TextInput
      name="title"
      label="Title"
      placeholder="React Fundamentals..."
      onChange={onChange}
      loading={loading}
      onFocus={onFocus}
      value={title}
      error={errors.title}
    />
    <SelectInput
      name="authorId"
      label="Author"
      defaultOption="Select Author"
      onChange={onChange}
      onFocus={onFocus}
      value={authorId}
      loading={loading}
      error={errors.authorId}
      options={authors}
    />
    <TextInput
      name="category"
      label="Category"
      placeholder="UI/UX, HTML, Graphql..."
      onChange={onChange}
      loading={loading}
      onFocus={onFocus}
      value={category}
      error={errors.category}
    />
    <TextInput
      name="length"
      label="Length"
      placeholder="5:12"
      onChange={onChange}
      onFocus={onFocus}
      loading={loading}
      value={length}
      error={errors.length}
    />
    <input
      type="submit"
      value={saving ? 'Saving...' : 'Save'}
      disabled={saving}
      className="btn btn-primary"
    />
  </form>
);

CourseForm.defaultProps = {
  title: '',
  authorId: '',
  category: '',
  length: '',
  errors: {},
  authors: []
};

CourseForm.propTypes = {
  title: string,
  authorId: string,
  category: string,
  length: string,
  errors: shape({}),
  authors: arrayOf(shape({})),
  saving: bool.isRequired,
  loading: bool.isRequired,
  onChange: func.isRequired,
  onSubmit: func.isRequired,
  onFocus: func.isRequired
};

export default CourseForm;
