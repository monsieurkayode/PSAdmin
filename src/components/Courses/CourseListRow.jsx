import React from 'react';
import { Link } from 'react-router-dom';
import { string, func } from 'prop-types';

/**
 *
 * @function CourseListRow
 *
 * @param {object} { id, watchHref, title, authorId, category, length }
 *
 * @returns {JSX}
 */
const CourseListRow = ({
  id, watchHref, title, authorId, category, length, deleteCourse
}) => (
  <tr>
    <td>
      <a href={watchHref} target="_blank" rel="noopener noreferrer">
        Watch
      </a>
    </td>
    <td><Link to={`/course/${id}`}>{title}</Link></td>
    <td>{authorId}</td>
    <td>{category}</td>
    <td>{length}</td>
    <td>
      <button
        type="button"
        className="btn btn-sm btn-danger"
        onClick={() => deleteCourse(id)}
      >
        Delete
      </button>
    </td>
  </tr>
);

CourseListRow.propTypes = {
  authorId: string.isRequired,
  category: string.isRequired,
  id: string.isRequired,
  length: string.isRequired,
  title: string.isRequired,
  watchHref: string.isRequired,
  deleteCourse: func.isRequired
};

export default CourseListRow;
