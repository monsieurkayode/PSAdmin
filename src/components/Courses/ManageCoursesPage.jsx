import React, { Component } from 'react';
import {
  shape, bool, arrayOf
} from 'prop-types';
import { connect } from 'react-redux';
import toastr from 'toastr';
import { Prompt } from 'react-router-dom';
import { saveCourse, setSaveErrors } from '../../actions/courseActions';
import CourseForm from './CourseForm';
import NotFoundPage from '../NotFound/NotFoundPage';
import { validateInput, checkChanges } from '../../helpers/validator';

import '../../styles/ManageCoursePage.scss';

/**
 * @class ManageCoursesPage
 *
 * @extends Component
 *
 * @returns {JSX}
 */
class ManageCoursesPage extends Component {
  static defaultProps = {
    course: {}
  }

  static propTypes = {
    isSaving: bool.isRequired,
    course: shape({}),
    authors: arrayOf(shape({})).isRequired,
    loadingCourse: bool.isRequired
  }

  static initialState = () => ({
    course: {
      id: '',
      title: '',
      length: '',
      authorId: '',
      category: '',
      watchHref: '',
    },
    errors: {},
    newChanges: false,
  });

  static getDerivedStateFromProps(
    { course, isSaving }, { errors, newChanges }
  ) {
    if (!isSaving && !newChanges) {
      return {
        ...ManageCoursesPage.initialState(),
        course,
        errors,
        newChanges
      };
    }

    return null;
  }

  state = ManageCoursesPage.initialState();

  redirect = this.props.history.push;

  componentWillUnmount() {
    this.props.setSaveErrors({});
  }

  /**
   * @method clearErrors
   *
   * @memberof ManageCoursesPage
   *
   * @returns {void}
   */
  clearErrors = () => {
    toastr.clear();
    if (Object.keys(this.state.errors).length > 0) {
      this.setState({ errors: {} });
    }
  }

  /**
   * @method handleSubmit
   *
   * @memberof ManageCoursesPage
   *
   * @returns {void}
   */
  handleSubmit = (event) => {
    event.preventDefault();
    const { course } = this.state;
    const { isValid, errors } = validateInput(course);

    this.setState({ newChanges: false });

    if (!isValid) {
      toastr.clear();
      toastr.error('Invalid form data', 'Submission Error!');
      return this.setState({ errors });
    }

    this.props.saveCourse(course)
      .then(() => this.redirect('/courses'))
      .catch(err => this.setState({ errors: err }));
  }

  /**
   * @method handleInputChange
   *
   * @memberof ManageCoursesPage
   *
   * @returns {void}
   */
  handleInputChange = (event) => {
    const { course } = this.state;
    course[event.target.name] = event.target.value;
    this.setState({ course, newChanges: checkChanges(course) });
  }

  render() {
    const {
      authors, isSaving, loadingCourse, match
    } = this.props;
    const { course, errors, newChanges } = this.state;

    if (!loadingCourse && match.params.id && !this.props.course.id) {
      return <NotFoundPage content="Course" />;
    }

    return (
      <div className="container" style={{ maxWidth: '55%', minWidth: '400px' }}>
        <Prompt
          when={!isSaving && newChanges}
          message="You have unsaved changes, leave page?"
        />
        <h1>Manage Course</h1>
        <CourseForm
          {...course}
          errors={errors}
          authors={authors}
          onChange={this.handleInputChange}
          onSubmit={this.handleSubmit}
          onFocus={this.clearErrors}
          saving={isSaving}
          loading={loadingCourse}
        />
      </div>
    );
  }
}

/**
 * @function getCourse
 *
 * @param {array} courses
 * @param {string} id
 *
 * @returns {any}
 */
const getCourse = (courses, id) => courses.find(c => c.id === id);

/**
 *
 * @description makes the ManageCoursePage a container component
 * by making it aware of the application state in the store
 * via the connect function from react-redux library
 *
 * @function mapStateToProps
 *
 * @param {object} { allCourses, allAuthors }
 */
const mapStateToProps = ({
  allAuthors: { authors },
  allCourses: {
    courses, isSaving, isLoading, errors
  }
}, { match: { params } }) => {
  let { course } = ManageCoursesPage.initialState();

  if (getCourse(courses, params.id)) {
    course = Object.assign({}, getCourse(courses, params.id));
  }

  const authorsFormattedForDropdown = authors.map(author => (
    {
      value: author.id,
      text: `${author.firstName} ${author.lastName}`
    }
  ));

  let loadingCourse = isLoading;

  if (!params.id) { loadingCourse = false; }

  return {
    isSaving,
    loadingCourse,
    errors,
    courses,
    course,
    authors: authorsFormattedForDropdown
  };
};

export default connect(
  mapStateToProps,
  { saveCourse, setSaveErrors }
)(ManageCoursesPage);
