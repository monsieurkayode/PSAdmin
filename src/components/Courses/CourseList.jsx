import React from 'react';
import { arrayOf, shape, func } from 'prop-types';
import CourseListRow from './CourseListRow';

/**
 *
 * @function CourseList
 *
 * @param {object} props
 *
 * @returns {JSX} JSX
 */
const CourseList = ({ courses, deleteCourse }) => (
  <table className="table">
    <thead>
      <tr>
        <th>&nbsp;</th>
        <th>Title</th>
        <th>Author</th>
        <th>Category</th>
        <th>Length</th>
        <th>&nbsp;</th>
      </tr>
    </thead>
    <tbody>
      {courses.map(course => (
        <CourseListRow
          key={course.id}
          {...course}
          deleteCourse={deleteCourse}
        />
      ))}
    </tbody>
  </table>
);

CourseList.propTypes = {
  courses: arrayOf(shape({})).isRequired,
  deleteCourse: func.isRequired
};

export default CourseList;
