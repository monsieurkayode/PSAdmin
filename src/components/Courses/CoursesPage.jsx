import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import {
  arrayOf, shape, bool, func
} from 'prop-types';
import { deleteCourse, pageChange } from '../../actions/courseActions';
import CourseList from './CourseList';
import Loader from '../common/Loader/Index';
import Pagination from '../common/Pagination/Index';
import paginator from '../../helpers/paginator';

/**
 *
 * @class CoursesPage
 * @extends Component
 *
 * @returns {JSX}
 */
class CoursesPage extends Component {
  componentWillUpdate({ meta, currentPage }) {
    if (meta.pageSize === 0 && currentPage > 1 && meta.pageCount >= 1) {
      this.props.pageChangeAction(currentPage - 1);
    }
  }

  /**
   * @method handlePageClick
   * @memberof CoursePage
   *
   * @param {object} { selected }
   *
   * @returns {void}
   */
  handlePageClick = ({ selected }) => {
    this.props.pageChangeAction(selected + 1);
  }

  render() {
    const {
      courses,
      isLoading,
      deleteCourseAction,
      meta
    } = this.props;
    return (
      <main>
        <div className="container jumbotron">
          <div className="container-fluid">
            <h1>Courses</h1>
            <p>Courses management page.</p>
            <Link to="/course">
              <button type="button" className="btn btn-lg btn-primary">
                Add Course
              </button>
            </Link>
          </div>
        </div>
        <section className="container">
          <div>
            {isLoading && <Loader size={40} />}
            {!isLoading && courses.length > 0
              && (
              <CourseList
                courses={courses}
                deleteCourse={deleteCourseAction}
              />
              )
            }
            {!isLoading && !courses.length
              && <div className="no-content">The course catalog is empty.</div>
            }
            {!isLoading && courses.length > 0 && (
            <Pagination
              {...meta}
              handlePageChange={this.handlePageClick}
            />
            )}
          </div>
        </section>
      </main>
    );
  }
}

CoursesPage.propTypes = {
  deleteCourseAction: func.isRequired,
  courses: arrayOf(shape({})).isRequired,
  isLoading: bool.isRequired,
  meta: shape({}).isRequired,
  pageChangeAction: func.isRequired
};

/**
 *
 * @description makes the CoursePage a container component
 * by making it aware of the application state in the store
 * via the connect function from react-redux library
 *
 * @function mapStateToProps
 *
 * @param {object} { allCourses }
 */
const mapStateToProps = ({
  allCourses: { courses, isLoading, currentPage }
}) => {
  const sortedCourses = courses.sort(
    (prev, next) => prev.title.toUpperCase() > next.title.toUpperCase()
  );
  const meta = paginator(sortedCourses, currentPage);

  return {
    courses: meta.results,
    isLoading,
    meta,
    currentPage
  };
};

export default connect(
  mapStateToProps,
  { deleteCourseAction: deleteCourse, pageChangeAction: pageChange }
)(CoursesPage);
