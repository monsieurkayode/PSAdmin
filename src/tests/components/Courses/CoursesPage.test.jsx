import React from 'react';
import { shallow } from 'enzyme';
import { CoursesPage } from '../../../components/Courses/CoursesPage';
import mockCourses from '../../__mocks__/mockCourses';

const setup = (isLoading, courses = []) => {
  const props = {
    isLoading,
    courses,
    deleteCourseAction: jest.fn(),
    pageChange: jest.fn(),
    currentPage: 2,
    meta: {
      pageSize: 0,
      pageCount: 2,
      page: 2
    }
  };

  const wrapper = shallow(<CoursesPage {...props} />);

  return {
    wrapper,
    props
  };
};

describe('CoursesPage Component', () => {
  it('renders correctly', () => {
    const { wrapper } = setup(false);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a main and a h1 tag element with Courses text', () => {
    const { wrapper } = setup(false);
    expect(wrapper.find('main').length).toBe(1);
    expect(wrapper.find('h1').length).toBe(1);
    expect(wrapper.find('h1').text()).toEqual('Courses');
  });

  it('renders Loader component when loading page', () => {
    const { wrapper } = setup(true);
    expect(wrapper.find('Loader').length).toBe(1);
  });

  describe('When the course list is empty', () => {
    const { wrapper } = setup(false);

    it('renders a no-content div', () => {
      const noContentDiv = wrapper.find('.no-content');
      expect(noContentDiv.length).toBe(1);
      expect(noContentDiv.text()).toEqual('The course catalog is empty.');
    });

    it('does not render a Pagination Component', () => {
      expect(wrapper.find('Pagination').length).toBeLessThan(1);
    });
  });

  describe('When the course list is not empty', () => {
    const { wrapper } = setup(false, mockCourses);
    it('renders a CourseList Component', () => {
      expect(wrapper.find('CourseList').length).toBe(1);
    });

    it('renders a Pagination Component', () => {
      expect(wrapper.find('Pagination').length).toBe(1);
    });
  });

  it('invokes componentWillUpdate when props and state changes', () => {
    const { wrapper, props: { meta, currentPage } } = setup(false, mockCourses);
    const componentWillUpdateSpy = jest.spyOn(
      wrapper.instance(), 'componentWillUpdate'
    );
    wrapper.instance().componentWillUpdate({ meta, currentPage });
    expect(componentWillUpdateSpy).toHaveBeenCalled();
  });

  it('invokes render method at least once', () => {
    const { wrapper } = setup(false, mockCourses);
    const renderSpy = jest.spyOn(wrapper.instance(), 'render');
    wrapper.instance().render();
    expect(renderSpy).toHaveBeenCalled();
  });

  it('invokes handlePageClick method', () => {
    const { wrapper } = setup(false, mockCourses);
    const handlePageClickSpy = jest.spyOn(
      wrapper.instance(), 'handlePageClick'
    );
    wrapper.instance().handlePageClick({ selected: 2 });
    expect(handlePageClickSpy).toHaveBeenCalled();
    expect(handlePageClickSpy).toHaveBeenCalledWith({ selected: 2 });
  });

  it('invoke componentWillUnmount when component is unmounting', () => {
    const { wrapper } = setup(false, mockCourses);
    const componentWillUnmountSpy = jest.spyOn(
      wrapper.instance(), 'componentWillUnmount'
    );
    wrapper.instance().componentWillUnmount();
    expect(componentWillUnmountSpy).toHaveBeenCalled();
  });
});
