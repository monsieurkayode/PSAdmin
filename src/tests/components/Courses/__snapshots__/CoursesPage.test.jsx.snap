// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoursesPage Component renders correctly 1`] = `
<main>
  <div
    className="container jumbotron"
  >
    <div
      className="container-fluid"
    >
      <h1>
        Courses
      </h1>
      <p>
        Courses management page.
      </p>
      <Link
        replace={false}
        to="/course"
      >
        <button
          className="btn btn-lg btn-primary"
          type="button"
        >
          Add Course
        </button>
      </Link>
    </div>
  </div>
  <section
    className="container"
  >
    <div>
      <div
        className="no-content"
      >
        The course catalog is empty.
      </div>
    </div>
  </section>
</main>
`;
